pkg/backends/influxdb/influxdb.go

```diff
@@ -0,0 +1,502 @@
+)
+
+// Client represents an InfluxDB client.
+type Client struct {
+	batchesCreated      uint64 // Accumulated number of batches created
+	batchesRetried      uint64 // Accumulated number of batches retried (first send is not a retry)
+	batchesDropped      uint64 // Accumulated number of batches aborted (data loss)
+	batchesCreateFailed uint64 // Accumulated number of batches which failed to serialize (data loss, no retry is possible)
+	batchesSent         uint64 // Accumulated number of batches successfully sent
+	seriesSent          uint64 // Accumulated number of series successfully sent
+
+	logger logrus.FieldLogger
+
+	credentials string
+	url         string
+
+	maxRequestElapsedTime time.Duration
+	client                *http.Client
+	metricsPerBatch       uint64
+	reqBufferSem          chan *bytes.Buffer // Two in one - a semaphore and a buffer pool
+	compressPayload       bool
+
+	disabledSubtypes gostatsd.TimerSubtypes
+	flushInterval    time.Duration
+}
+
+// NewClientFromViper returns a new InfluxDB API client.
+func NewClientFromViper(
+	v *viper.Viper,
+	logger logrus.FieldLogger,
+	pool *transport.TransportPool,
+) (gostatsd.Backend, error) {
+	influxViper := util.GetSubViper(v, "influxdb")
+	influxViper.SetDefault(paramApiEndpoint, "")
+	influxViper.SetDefault(paramCompressPayload, true)
+	influxViper.SetDefault(paramCredentials, "")
+	influxViper.SetDefault(paramMaxRequestElapsedTime, defaultMaxRequestElapsedTime)
+	influxViper.SetDefault(paramMaxRequests, defaultMaxRequests)
+	influxViper.SetDefault(paramMetricsPerBatch, defaultMetricsPerBatch)
+	influxViper.SetDefault(paramTransport, "default")
+
+	cfg, err := newConfigFromViper(influxViper, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewClient(
+		influxViper.GetString(paramApiEndpoint),
+		influxViper.GetBool(paramCompressPayload),
+		influxViper.GetString(paramCredentials),
+		influxViper.GetUint(paramMaxRequests),
+		influxViper.GetDuration(paramMaxRequestElapsedTime),
+		influxViper.GetUint64(paramMetricsPerBatch),
+		influxViper.GetString(paramTransport),
+		cfg,
+		gostatsd.DisabledSubMetrics(v),
+		logger,
+		pool,
+	)
+}
+
+// NewClient returns a new InfluxDB API client.
+func NewClient(
+	apiEndpoint string,
+	compressPayload bool,
+	credentials string,
+	maxRequests uint,
+	maxRequestElapsedTime time.Duration,
+	metricsPerBatch uint64,
+	transport string,
+	cfg config,
+	disabled gostatsd.TimerSubtypes,
+	logger logrus.FieldLogger,
+	pool *transport.TransportPool,
+) (*Client, error) {
+
+	if apiEndpoint == "" {
+		return nil, errApiEndpointRequired
+	}
+	if maxRequests == 0 {
+		return nil, errMaxRequestsIsNotPositive
+	}
+	if maxRequestElapsedTime <= 0 && maxRequestElapsedTime != -1 {
+		return nil, errMaxRequestElapsedTimeInvalid
+	}
+	if metricsPerBatch == 0 {
+		return nil, errMetricsPerBatchIsNotPositive
+	}
+	parsedEndpoint, err := url.Parse(apiEndpoint)
```

Comment:
Is it worthwhile checking the scheme to set with the url?
assuming it has to be `http` or `https`?

Reply:
Nah. If you want to post to an ftp site, then good luck!  It's up to the http.Client to cleanup that mess, not us.

Reply:
For what it's worth, InfluxDB apparently supports ingestion via http(s) over unix domain socket. I have no idea if this will successfully talk to it, but it probably would, since they're both go, so they should have compatible notions of how to talk over it.